### PR TITLE
Add support for building arm64 images using local envoy.

### PR DESCRIPTION
**Pull request by jjjighg**
We sucessfully build with this command:
USE_LOCAL_PROXY=1 make docker.base BAZEL_BUILD_ARGS="--cpu aarch64"
docker tag istio/base:${TAG} istio/base:${BASE_VERSION}
USE_LOCAL_PROXY=1 make docker BAZEL_BUILD_ARGS="--cpu aarch64"

### CODE REVIEW

**Comment by kebe7jun** [comment]
Awesome, we also need it.
/ok-to-test

**Comment by hzxuzhonghu** [comment]
@howardjohn For arm support, i think it is reasonable.

**Comment by hzxuzhonghu** [comment]
There is also a pr to build envoy for arm https://github.com/envoyproxy/envoy-wasm/pull/357

File: docker/Dockerfile.istiod [hunk 1]
```diff
@@ -11,7 +11,7 @@ USER 1337:1337
 
 # The following section is used as base image if BASE_DISTRIBUTION=distroless
 # hadolint ignore=DL3007
-FROM gcr.io/distroless/static:nonroot as distroless
+FROM discolix/static:nonroot as distroless
```
Comment (brian-avery) [inline docker/Dockerfile.istiod]: was this change intentional? I don't see discolix defined on Docker Hub or grc.io

Comment (brian-avery) [reply docker/Dockerfile.istiod]: Nevermind. Found it on Dockerhub. 
https://hub.docker.com/r/discolix/static

File: docker/Dockerfile.kubectl [hunk 2]
```diff
@@ -6,7 +6,8 @@ ARG BASE_VERSION=latest
 FROM docker.io/istio/base:${BASE_VERSION} as default
```
Comment (sdake) [inline docker/Dockerfile.kubectl]: I believe in master we no longer use this image after the deletion of the `installer` directory.

File: docker/Dockerfile.istiod (hunk 1 above)
Comment (sdake) [reply docker/Dockerfile.istiod]: I would prefer to see distroless divided by architecture, or even better for gcr.io/distroless/static to support arm targets. The discolix image is nearly 2 months old as of this review.

File: Makefile.core.mk [hunk 3]
```diff
@@ -296,7 +296,7 @@ ifeq ($(BUILD_ALL),true)
 $(ISTIO_OUT_LINUX)/$(shell basename $(1)): build-linux
 else
 $(ISTIO_OUT_LINUX)/$(shell basename $(1)): $(ISTIO_OUT_LINUX)
-	STATIC=0 GOOS=linux GOARCH=amd64 LDFLAGS='-extldflags -static -s -w' common/scripts/gobuild.sh $(ISTIO_OUT_LINUX)/ $(1)
+	STATIC=0 GOOS=linux GOARCH=$(GOARCH_LOCAL) LDFLAGS='-extldflags -static -s -w' common/scripts/gobuild.sh $(ISTIO_OUT_LINUX)/ $(1)
```
Comment (sdake) [inline Makefile.core.mk]: This isn't correct. The build system will need significant modification to support arm. Keep in mind the development environment cannot see any regression. This would break the 60 or so test cases for the developer system (that just merged btw..) to enable arm. If you need more information, feel free to check out the developer wiki: https://github.com/istio/istio/wiki

File: Makefile [hunk 4]
```diff
@@ -53,11 +53,11 @@ else
 endif
 
 export TARGET_OUT ?= $(shell pwd)/out/$(TARGET_OS)_$(TARGET_ARCH)
-export TARGET_OUT_LINUX ?= $(shell pwd)/out/linux_amd64
+export TARGET_OUT_LINUX ?= $(shell pwd)/out/linux_$(TARGET_ARCH)
```
Comment (sdake) [inline Makefile]: Top-level makefile changes should be changed in https://github.com/istio/common-files. This change would be overwritten in any update of the common-files repository.

Also this change is incorrect. TARGET_OUT_LINUX should be generalized - see https://github.com/istio/istio/blob/master/Makefile.core.mk#L290-L293 for directional guidance on what needs to be done here.

File: galley/docker/Dockerfile.galley [hunk 5]
```diff
@@ -11,7 +12,7 @@ USER 1337:1337
 
 # The following section is used as base image if BASE_DISTRIBUTION=distroless
 # hadolint ignore=DL3007
-FROM gcr.io/distroless/static:nonroot as distroless
+FROM ${DISTROLESS_IMAGE}/static:nonroot as distroless
```
Comment (sdake) [inline galley/docker/Dockerfile.galley]: I could see how this could work. This is a better approach than that used in Dockerfile.istiod above.

File: tools/istio-docker.mk [hunk 6]
```diff
@@ -17,6 +17,7 @@
 .PHONY: docker.save
 .PHONY: docker.push
 
+
```
Comment (sdake) [inline tools/istio-docker.mk]: please remove unnecessary whitespace change.

File: tools/istio-docker.mk [hunk 7]
```diff
@@ -27,6 +28,14 @@ DOCKER_TARGETS ?= docker.pilot docker.proxytproxy docker.proxyv2 docker.app dock
 	docker.mixer docker.mixer_codegen docker.citadel docker.galley docker.sidecar_injector docker.kubectl docker.node-agent-k8s \
 	docker.istioctl
 
+ifeq ($(TARGET_ARCH),amd64)
+	DOCKER_OPT=--build-arg DISTROLESS_IMAGE=gcr.io/distroless
+else ifeq ($(TARGET_ARCH),arm64)
+	DOCKER_OPT=--build-arg DISTROLESS_IMAGE=discolix
```
Comment (sdake) [inline tools/istio-docker.mk]: DISTROLESS_IMAGE=`docker.io/discolix`. This is necessary as the CRI in use may not always use docker as a default. This project always encodes the HUB in every image.

File: tools/istio-docker.mk [hunk 8]
```diff
@@ -27,6 +28,14 @@ DOCKER_TARGETS ?= docker.pilot docker.proxytproxy docker.proxyv2 docker.app dock
...
+else
+	DOCKER_OPT=--build-arg DISTROLESS_IMAGE=gcr.io/distroless
```
Comment (sdake) [inline tools/istio-docker.mk]: This else conditional should be an error statement rather than a repeat of line 32.

File: tools/istio-docker.mk [hunk 9]
```diff
@@ -76,7 +85,7 @@ $(foreach FILE,$(DOCKER_FILES_FROM_ISTIO_BIN), \
         $(eval $(ISTIO_DOCKER)/$(FILE): $(ISTIO_BIN)/$(FILE) | $(ISTIO_DOCKER); cp $(ISTIO_BIN)/$(FILE) $(ISTIO_DOCKER)/$(FILE)))
 
 docker.sidecar_injector: BUILD_PRE=&& chmod 755 sidecar-injector
-docker.sidecar_injector: BUILD_ARGS=--build-arg BASE_VERSION=${BASE_VERSION}
+docker.sidecar_injector: BUILD_ARGS=--build-arg BASE_VERSION=${BASE_VERSION} ${DOCKER_OPT}
```
Comment (sdake) [inline tools/istio-docker.mk]: all these changes don't feel right. I think there is a better way to do this by encoding `DOCKER_OPT` (which should probably be renamed to `BASE_IAMGE` in the makefile rule on line 285 of this file.

File: tools/istio-docker.mk [hunk 10]
```diff
@@ -238,12 +247,12 @@ docker.citadel-test: $(ISTIO_DOCKER)/istio_ca.crt
...
 	$(DOCKER_RULE)
 
-docker.node-agent: BUILD_ARGS=--build-arg BASE_VERSION=${BASE_VERSION}
+docker.node-agent: BUILD_ARGS=--build-arg BASE_VERSION=${BASE_VERSION} ${DOCKER_OPT}
 docker.node-agent: security/docker/Dockerfile.node-agent
 docker.node-agent: $(ISTIO_DOCKER)/node_agent
 	$(DOCKER_RULE)
 
-docker.node-agent-k8s: BUILD_ARGS=--build-arg BASE_VERSION=${BASE_VERSION}
+docker.node-agent-k8s: BUILD_ARGS=--build-arg BASE_VERSION=${BASE_VERSION} ${DOCKER_OPT}
```
Comment (sdake) [inline tools/istio-docker.mk]: All these extra DOCKER_OPTS may be good for a prototype. They are not helpful for long term maintenance - and instead this BASE_OPT should be integrated into line 285.

**Comment by sdake** [comment]
I see this PR was marked approved by @hzxuzhonghu. I think this PR needs more development time, so I have marked it `work-in-progress` until the technical issues can be sorted out.

Cheers
-steve

**Comment by hzxuzhonghu** [comment]
@jjjighg Are you still working on this?
